src/torrent.cpp

```diff
@@ -5678,61 +5680,81 @@ namespace {
 
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
 
-		if (j->ret == piece_manager::no_error)
+		if (m_file_priority != *p)
 		{
-			TORRENT_ASSERT(!j->error);
 			m_file_priority = *p;
 			update_piece_priorities();
 		}
-		else
-		{
-			// in this case, some file priorities failed to get set
-			TORRENT_ASSERT(j->error);
-			TORRENT_ASSERT(!m_share_mode);
 
-			if (alerts().should_post<file_error_alert>())
-				alerts().emplace_alert<file_error_alert>(j->error.ec
-					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+		if (!j->error) return;
 
-			set_error(j->error.ec, j->error.file);
-			pause();
-		}
+		// in this case, some file priorities failed to get set
+		TORRENT_ASSERT(j->error);
+		TORRENT_ASSERT(!m_share_mode);
+
+		if (alerts().should_post<file_error_alert>())
+			alerts().emplace_alert<file_error_alert>(j->error.ec
+				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+
+		set_error(j->error.ec, j->error.file);
+		pause();
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
+		if (is_seed()) return;
+
 		// store file priorities until we get the metadata
 		if (!valid_metadata())
 		{
 			m_file_priority.assign(files.begin(), files.end());
 			return;
 		}
 
+		// the vector need to have exactly one element for every file
+		// in the torrent
+		TORRENT_ASSERT(int(files.size()) == m_torrent_file->num_files());
+
 		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
 
-		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
+		if (new_priority == m_file_priority)
+		{
+			return;
+		}
 
 		// storage may be NULL during shutdown
-		if (boost::shared_ptr<piece_manager> const tmp = m_storage)
+		if (m_torrent_file->num_pieces() > 0 && m_storage)
```

Comment:
How is it ever possible, that torrent has file(s) and no pieces?

Reply:
I think this may be some left-over from a feature where the bulk of the metadata for a torrent could be "evicted" if the torrent was inactive. It should be fixed, but I would prefer to keep the changes separate. And I prefer to make code quality improvements in master, to avoid excessive merging

Reply:
If I get it right, by 'evicting' you mean torrent load/unload feature. If so, the torrent that has been unloaded will have `m_num_pieces` value unchanged.

Reply:
yeah, I also noticed that calling `num_pieces()` when we don't have metadata is an error, so as part of simplifying this function (and make the metadata and no-metadata paths more similar) I removed this again.